File: pathfinding algorithm/Astar.py
// note, chatgpt and geeksforgeeks was used to help with understanding this algorithm although i would say that my
// understanding still isnt the best
import heapq

# Heuristic function: This function estimates the cost to reach the goal from node 'a'.
# Here, it uses the Manhattan distance (good for grids where you can only move in 4 directions).
def heuristic(a, b):
    return abs(a[0] - b[0]) + abs(a[1] - b[1])

# A* search algorithm to find the shortest path in a maze from 'start' to 'end'.
def algorithm(maze, start, end):
    neighbours = [(0, 1), (1, 0), (0, -1), (-1, 0)]  # Possible moves: right, down, left, up
    close_set = set()  # Set of nodes that have already been evaluated
    came_from = {}  # Dictionary to track the best previous node for each node
    gscore = {start: 0}  # Dictionary to track the cost from start to a node
    fscore = {start: heuristic(start, end)}  # Dictionary to track the estimated total cost from start to end
    oheap = []  # Priority queue (min-heap) to select the next node to explore
    
    # Start by pushing the starting node into the heap, with its fscore as the priority
    heapq.heappush(oheap, (fscore[start], start))
    
    # Main loop: keep processing nodes until there are none left to explore
    while oheap:
        # Pop the node with the lowest fscore from the heap (i.e., the most promising node)
        current = heapq.heappop(oheap)[1]
        
        # If we've reached the end, reconstruct the path by tracing back through 'came_from'
        if current == end:
            data = []  # List to store the path
            while current in came_from:  # Keep moving backwards through the path
                data.append(current)
                current = came_from[current]
            data.append(start)  # Don't forget to add the start node
            return data[::-1]  # Reverse the path to get it from start to end
        
        close_set.add(current)  # Mark the current node as evaluated (closed)
        
        # Explore each of the neighboring nodes
        for i, j in neighbours:
            neighbour = current[0] + i, current[1] + j  # Calculate the neighbor's position
            tentative_g_score = gscore[current] + 1  # Calculate the cost to reach this neighbor
            
            # Check if the neighbor is within the maze bounds
            if 0 <= neighbour[0] < len(maze[0]):
                if 0 <= neighbour[1] < len(maze):
                    if maze[neighbour[1]][neighbour[0]] == 1:  # Check if the neighbor is a wall
                        continue  # Skip if it's a wall
                else:
                    continue  # Skip if the neighbor is out of bounds vertically
            else:
                continue  # Skip if the neighbor is out of bounds horizontally
                
            # If this path is not better than any previously found, skip it
            if neighbour in close_set and tentative_g_score >= gscore.get(neighbour, 0):
                continue
                
            # If this path is better, or the neighbor hasn't been explored, update it
            if tentative_g_score < gscore.get(neighbour, 0) or neighbour not in [i[1] for i in oheap]:
                came_from[neighbour] = current  # Record the best previous node
                gscore[neighbour] = tentative_g_score  # Update the gscore for this node
                fscore[neighbour] = tentative_g_score + heuristic(neighbour, end)  # Update the fscore
                heapq.heappush(oheap, (fscore[neighbour], neighbour))  # Push the neighbor to the heap
                
    return False  # If no path is found, return False

File: pathfinding algorithm/Astar.c
// note, chatgpt and geeksforgeeks was used to help with understanding this algorithm although i would say that my
// understanding still isnt the best
#include "astar.h"
#include <stdio.h>
#include <stdlib.h>
#include <limits.h>
#include <stdbool.h>

#define MAX_QUEUE_SIZE 1000

// used to refer to a point in the maze
typedef struct {
    int x, y;
} Point;

// Structure to represent a node in the A* algorithm
typedef struct {
    Point point; // The position of the node
    int g; // Cost from the start node to this node
    int h; // Heuristic cost estimate to the goal
    int f; // Total cost (g + h)
} Node;

// Priority queue is used to explore nodes
typedef struct {
    Node nodes[MAX_QUEUE_SIZE];
    int size;
} PriorityQueue;

// Function to swap two nodes in the priority queue
static void swap(Node* a, Node* b) {
    Node temp = *a;
    *a = *b;
    *b = temp;
}

// this is used to put a node to the priority queue
static void push(PriorityQueue* pq, Node node) {
    int i = pq->size++;
    // Place the new node in its correct position based on its f-cost
    while (i > 0 && node.f < pq->nodes[(i - 1) / 2].f) {
        pq->nodes[i] = pq->nodes[(i - 1) / 2];
        i = (i - 1) / 2;
    }
    pq->nodes[i] = node;
}

// Function to pop the node with the lowest f-cost from the priority queue
static Node pop(PriorityQueue* pq) {
    Node root = pq->nodes[0];
    Node last = pq->nodes[--pq->size];
    int i = 0;
    int child;
    // Reheapify the priority queue
    while ((child = 2 * i + 1) < pq->size) {
        if (child + 1 < pq->size && pq->nodes[child + 1].f < pq->nodes[child].f) {
            child++;
        }
        if (last.f <= pq->nodes[child].f) break;
        pq->nodes[i] = pq->nodes[child];
        i = child;
    }
    pq->nodes[i] = last;
    return root;
}

static bool is_valid_move(int x, int y, unsigned char map[15][19], bool visited[15][19]) {
    return x >= 0 && x < 19 && y >= 0 && y < 15 && map[y][x] != 1 && !visited[y][x];
}

// Function to calculate the heuristic (Manhattan distance) between two points
static int heuristic(int x1, int y1, int x2, int y2) {
    return abs(x1 - x2) + abs(y1 - y2);
}

// used to update the map
static void print_path(Point prev[15][19], Point start, Point end, unsigned char path[15][19]) {
    Point current = end;
    while (current.x != start.x || current.y != start.y) {
        path[current.y][current.x] = 9; 
        current = prev[current.y][current.x];
    }
    path[start.y][start.x] = 2; 
}

void astar(unsigned char map[15][19], unsigned char start_x, unsigned char start_y, unsigned char dest_x, unsigned char dest_y, unsigned char path[15][19], int *shortest_length) {
    PriorityQueue open_list = {{0}, 0}; // Initialize the priority queue
    bool visited[15][19]; // Array to keep track of visited nodes
    Point prev[15][19]; // Array to keep track of the previous node for each node (for path reconstruction)
    // possible x and y movements
    int dx[] = {0, 0, -1, 1}; 
    int dy[] = {-1, 1, 0, 0}; 
    
    // Initialize the visited and prev arrays
    for (int y = 0; y < 15; y++) {
        for (int x = 0; x < 19; x++) {
            visited[y][x] = false;
            prev[y][x] = (Point){-1, -1};
        }
    }

    // Create and push the start node onto the priority queue
    Node start_node = { {start_x, start_y}, 0, heuristic(start_x, start_y, dest_x, dest_y), 0 };
    push(&open_list, start_node);
    visited[start_y][start_x] = true;

    // Main loop to explore the nodes
    while (open_list.size > 0) {
        Node current_node = pop(&open_list); // Get the node with the lowest f-cost
        Point u = current_node.point;

        // if we have reached the end node, updated the shortest path
        if (u.x == dest_x && u.y == dest_y) {
            *shortest_length = current_node.g;
            break;
        }

        // Explore the neighbors of the current node
        for (int i = 0; i < 4; i++) {
            int new_x = u.x + dx[i];
            int new_y = u.y + dy[i];

            // If the move is valid, calculate the costs and push the neighbor onto the priority queue
            if (is_valid_move(new_x, new_y, map, visited)) {
                int new_g = current_node.g + 1; // Cost from start to this neighbor
                int new_h = heuristic(new_x, new_y, dest_x, dest_y); // Heuristic cost to goal
                int new_f = new_g + new_h; // Total cost

                Node neighbour_node = { {new_x, new_y}, new_g, new_h, new_f };
                if (!visited[new_y][new_x]) {
                    visited[new_y][new_x] = true; // Mark the neighbor as visited
                    prev[new_y][new_x] = u; // Set the previous node for path reconstruction
                    push(&open_list, neighbour_node); // Push the neighbor onto the priority queue
                }
            }
        }
    }

    // copies the map into path
    for (int y = 0; y < 15; y++) {
        for (int x = 0; x < 19; x++) {
            path[y][x] = map[y][x];
        }
    }

    print_path(prev, (Point){start_x, start_y}, (Point){dest_x, dest_y}, path);
}
